### PR TITLE
[WIP] Extend generator trait to return batches of rows at the same logical offset

### PR DESCRIPTION
**Pull request by sjwiesman**
### Motivation

Strict serializability is an important feature that is often difficult to explain to users who have not felt the pain in other solutions. It is more easily demonstratable when definining views across multiple input sources that are modified within the same transaction, and with more complex schemas. This change extends the Generator trait to return batches of rows that are all produced at the same logical offset. The second commit then extends the auctions source with additional views whose semantics require transactional updates to provide consistent results.   

### Tips for reviewer

This is the first time I've tried to work in the materialize code base. Do not hesitate to give me any and all feedback :) This is in no way time sensitive. 

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

File: src/storage/src/source/generator/auction.rs [hunk 1]
```diff
@@ -55,13 +58,41 @@ impl Generator for Auction {
...
                     .with_column("bid_time", ScalarType::TimestampTz.nullable(false)),
             ),
+            (
+                "customers",
+                RelationDesc::empty()
+                    .with_column("id", ScalarType::Int64.nullable(false))
+                    .with_column("name", ScalarType::String.nullable(false)),
+            ),
+            (
+                "customer_bids",
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply src/storage/src/source/generator/auction.rs:65]: You are right, I need to think through this more. I want to show off more complex joins.

File: src/storage/src/source/generator.rs [hunk 2]
```diff
@@ -78,42 +79,86 @@ impl SourceReader for LoadGeneratorSourceReader {
...
-            offset,
+            between_batches: true,
         })
     }
 
     fn get_next_message(
         &mut self,
     ) -> Result<NextMessage<Self::Key, Self::Value, Self::Diff>, SourceReaderError> {
-        if self.last.elapsed() < self.tick {
+        if self.between_batches && self.last.elapsed() < self.tick {
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply src/storage/src/source/generator.rs]: This was the part I was unsure of and wanted to ask you about. I'll defer to your judgement here, I don't really know what tick does

File: src/storage/src/source/generator.rs [hunk 3]
```diff
@@ -78,42 +79,86 @@ impl SourceReader for LoadGeneratorSourceReader {
...
         self.last += self.tick;
-        self.offset += 1;
         match self.rows.next() {
-            Some(value) => Ok(NextMessage::Ready(SourceMessageType::Finalized(
-                SourceMessage {
+            Some(value) => {
+                self.between_batches = match value {
+                    SourceMessageType::Finalized(_) => true,
+                    _ => false,
+                };
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply src/storage/src/source/generator.rs]: well that's better...

**Comment by sjwiesman** [comment]
We can now demonstrate consistency via account transfers based on winning bids

```sql
CREATE VIEW account_balances 
AS 
  WITH initial_balance 
       AS (SELECT id AS account_id, 
                  balance 
           FROM   accounts), 
       credits 
       AS (SELECT accounts.id AS account_id, 
                  Sum(amount) AS credit 
           FROM   winning_bids 
                  JOIN users 
                    ON winning_bids.seller = users.id 
                  JOIN accounts 
                    ON users.org_id = accounts.org_id 
           GROUP  BY accounts.id), 
       debits 
       AS (SELECT accounts.id AS account_id, 
                  Sum(amount) AS debit 
           FROM   winning_bids 
                  JOIN users 
                    ON winning_bids.buyer = users.id 
                  JOIN accounts 
                    ON users.org_id = accounts.org_id 
           GROUP  BY accounts.id) 
  SELECT initial_balance.account_id, 
         Sum(initial_balance.balance) 
         + Sum(credits.credit) - Sum(debits.debit) AS balance 
  FROM   initial_balance, 
         credits, 
         debits 
  WHERE  initial_balance.account_id = credits.account_id 
         AND initial_balance.account_id = debits.account_id 
  GROUP  BY initial_balance.account_id;
```

You can verify consistency by showing credits - debits is always 0. No money is ever created or destroyed. 

```sql
COPY (
    TAIL (
         with credits as (SELECT accounts.id AS account_id,
                  Sum(amount) AS credit
           FROM   winning_bids
                  JOIN users
                    ON winning_bids.seller = users.id
                  JOIN accounts
                    ON users.org_id = accounts.org_id
           GROUP  BY accounts.id
      ), debits as (
      SELECT accounts.id AS account_id,
                        Sum(amount) AS debit
                 FROM   winning_bids
                        JOIN users
                          ON winning_bids.buyer = users.id
                        JOIN accounts
                          ON users.org_id = accounts.org_id
                 GROUP  BY accounts.id) select sum(credit) - sum(debit) from credits, debit
    )
) TO STDOUT
```

**Review by mjibson — APPROVED** [review]

**Comment by mjibson** [comment]
You'll have to add the new tables to the test output.

**Pull request merged**